File: apollo-api/src/main/java/com/apollographql/apollo/api/InputFieldWriter.java
package com.apollographql.apollo.api;

import java.io.IOException;
import java.util.Map;

import org.jetbrains.annotations.NotNull;

public interface InputFieldWriter {
  void writeString(@NotNull String fieldName, String value) throws IOException;

  void writeInt(@NotNull String fieldName, Integer value) throws IOException;

  void writeLong(@NotNull String fieldName, Long value) throws IOException;

  void writeDouble(@NotNull String fieldName, Double value) throws IOException;

  void writeNumber(@NotNull String fieldName, Number value) throws IOException;

  void writeBoolean(@NotNull String fieldName, Boolean value) throws IOException;

  void writeCustom(@NotNull String fieldName, ScalarType scalarType, Object value) throws IOException;

  void writeObject(@NotNull String fieldName, InputFieldMarshaller marshaller) throws IOException;

  void writeList(@NotNull String fieldName, ListWriter listWriter) throws IOException;

  void writeMap(@NotNull String fieldName, Map<String, Object> value) throws IOException;

  interface ListWriter {
    void write(@NotNull ListItemWriter listItemWriter) throws IOException;
  }

  interface ListItemWriter {
    void writeString(String value) throws IOException;

    void writeInt(Integer value) throws IOException;

    void writeLong(Long value) throws IOException;

    void writeDouble(Double value) throws IOException;

    void writeNumber(Number value) throws IOException;

    void writeBoolean(Boolean value) throws IOException;

    void writeCustom(ScalarType scalarType, Object value) throws IOException;

    void writeObject(InputFieldMarshaller marshaller) throws IOException;

    void writeList(ListWriter listWriter) throws IOException;

    void writeMap(Map<String, Object> value) throws IOException;
  }
}

File: apollo-api/src/main/java/com/apollographql/apollo/api/InputFieldWriter.kt
package com.apollographql.apollo.api

import java.io.IOException

interface InputFieldWriter {
  @Throws(IOException::class)
  fun writeString(fieldName: String, value: String?)

  @Throws(IOException::class)
  fun writeInt(fieldName: String, value: Int?)

  @Throws(IOException::class)
  fun writeLong(fieldName: String, value: Long?)

  @Throws(IOException::class)
  fun writeDouble(fieldName: String, value: Double?)

  @Throws(IOException::class)
  fun writeNumber(fieldName: String, value: Number?)

  @Throws(IOException::class)
  fun writeBoolean(fieldName: String, value: Boolean?)

  @Throws(IOException::class)
  fun writeCustom(fieldName: String, scalarType: ScalarType, value: Any?)

  @Throws(IOException::class)
  fun writeObject(fieldName: String, marshaller: InputFieldMarshaller?)

  @Throws(IOException::class)
  fun writeList(fieldName: String, listWriter: ListWriter?)

  @Throws(IOException::class)
  fun writeMap(fieldName: String, value: Map<String, @JvmSuppressWildcards Any>?)

  interface ListWriter {
    @Throws(IOException::class)
    fun write(listItemWriter: ListItemWriter)
  }

  interface ListItemWriter {
    @Throws(IOException::class)
    fun writeString(value: String?)

    @Throws(IOException::class)
    fun writeInt(value: Int?)

    @Throws(IOException::class)
    fun writeLong(value: Long?)

    @Throws(IOException::class)
    fun writeDouble(value: Double?)

    @Throws(IOException::class)
    fun writeNumber(value: Number?)

    @Throws(IOException::class)
    fun writeBoolean(value: Boolean?)

    @Throws(IOException::class)
    fun writeCustom(scalarType: ScalarType, value: Any?)

    @Throws(IOException::class)
    fun writeObject(marshaller: InputFieldMarshaller?)

    @Throws(IOException::class)
    fun writeList(listWriter: ListWriter?)

    @Throws(IOException::class)
    fun writeMap(value: Map<String, @JvmSuppressWildcards Any>?)
  }
}
